File: src/ladym/langgraph/_runtime.py
"""Shared runtime helpers for the LangGraph integration (paths A and B).

Keeps Engine lifecycle and (for the nodes path) workspace resolution in one
place so the two paths stay consistent and have a single seam against Engine.
"""
from __future__ import annotations

import copy
from pathlib import Path
from typing import Any

from ..config import Config
from ..engine import Engine


def resolve_engine(
    engine: Engine | Config | str | Path | None,
    *,
    workspace: str | None = None,
) -> Engine:
    """Return a ready-to-use Engine from flexible factory input.

    * ``Engine`` -> returned as-is (caller owns lifecycle); ``workspace`` is
      IGNORED — the engine's own ``config.workspace`` wins. Pass a ``Config``
      or path instead if you need a different workspace.
    * ``Config`` -> new Engine from a copy; ``workspace`` overrides ``cfg.workspace``.
    * ``str`` / ``Path`` -> ``Config(db_path=...)``; ``workspace`` honored.
    * ``None`` -> default ``Config()`` (offline hashing embedding); ``workspace`` honored.
    """
    if isinstance(engine, Engine):
        return engine
    if isinstance(engine, Config):
        cfg = copy.copy(engine)
    elif isinstance(engine, (str, Path)):
        cfg = Config(db_path=Path(engine))
    else:
        cfg = Config()
    if workspace:
        cfg.workspace = workspace
    return Engine(cfg)


def resolve_workspace(config: Any, engine: Engine) -> str:
    """Resolve the ladyM workspace for a LangGraph node invocation (path B).

    Reads ``config["configurable"]["user_id"]``; falls back to
    ``engine.config.workspace``. Typed ``Any`` so this module has no hard
    langchain dependency.
    """
    try:
        configurable = (config or {}).get("configurable", {}) or {}
    except AttributeError:
        return engine.config.workspace
    user_id = configurable.get("user_id")
    return str(user_id) if user_id else engine.config.workspace

File: langgraph/runtime.go
// Shared runtime helpers for the LangGraph integration (paths A and B),
// mirroring Python's langgraph/_runtime.py on the main branch. Keeps Engine
// lifecycle and workspace resolution in one place so the two paths stay
// consistent and have a single seam against Engine.
package langgraph

import (
	"fmt"

	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/engine"
	lgruntime "github.com/projanvil/langchain-golang/langgraph/runtime"
)

// ResolveEngine returns a ready-to-use Engine from flexible factory input,
// mirroring Python's resolve_engine:
//
//   - *engine.Engine — returned as-is (caller owns lifecycle); workspace is
//     IGNORED — the engine's own Config.Workspace wins. Pass a *config.Config
//     or db path instead if you need a different workspace.
//   - *config.Config — new Engine from a shallow copy; a non-empty workspace
//     overrides cfg.Workspace.
//   - string — treated as a db path over config.Default(); workspace honored.
//   - nil — config.Default() (offline hashing embedding); workspace honored.
//
// Any other type is an error.
func ResolveEngine(src any, workspace string) (*engine.Engine, error) {
	switch v := src.(type) {
	case *engine.Engine:
		return v, nil
	case *config.Config:
		cfg := *v // shallow copy, retarget workspace
		if workspace != "" {
			cfg.Workspace = workspace
		}
		return engine.New(&cfg)
	case string:
		cfg := config.Default()
		cfg.DBPath = v
		if workspace != "" {
			cfg.Workspace = workspace
		}
		return engine.New(cfg)
	case nil:
		cfg := config.Default()
		if workspace != "" {
			cfg.Workspace = workspace
		}
		return engine.New(cfg)
	default:
		return nil, fmt.Errorf("langgraph: unsupported engine source %T (want *engine.Engine, *config.Config, db path string, or nil)", src)
	}
}

// WorkspaceFromUserID returns a WorkspaceFunc that resolves the ladyM
// workspace from the run-scoped context's "user_id" field, mirroring
// Python's resolve_workspace reading config["configurable"]["user_id"].
// An empty result falls back to the engine's default workspace.
func WorkspaceFromUserID() WorkspaceFunc {
	return func(rt lgruntime.Runtime) string {
		if v, ok := lgruntime.ValueFromRuntime(rt, "user_id"); ok {
			if s, ok := v.(string); ok {
				return s
			}
		}
		return ""
	}
}
